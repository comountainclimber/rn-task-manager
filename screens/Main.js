import React from "react";
import {
  StyleSheet,
  ScrollView,
  KeyboardAvoidingView,
  Keyboard,
  Dimensions,
  View,
  SafeAreaView
} from "react-native";
import moment from "moment";

import { DATE, TASKS } from "../constants";
import Calendar from "../components/Calendar";
import CurrentDayButton from "../components/CurrentDayButton";
import TaskInput from "../components/TaskInput";
import SwipeableTaskList from "../components/SwipeableTaskList";

export default function Main() {
  const [selectedDate, setSelectedDate] = React.useState(
    moment(DATE).format("YYYY-MM-DD")
  );
  const [tasks, setTasks] = React.useState(TASKS);
  const [calendarVisible, toggleCalendarVisibility] = React.useState(false);
  const [taskRowBeingMoved, setTaskRowBeingMoved] = React.useState(null);

  const handleDayPress = date => setSelectedDate(date.dateString);

  const handleAddTask = ({ value }) => {
    const copy = { ...tasks };

    if (!copy[selectedDate]) {
      copy[selectedDate] = [];
    }

    copy[selectedDate].unshift({
      completed: false,
      description: value,
      id: String(Date.now())
    });
    setTasks(copy);
    Keyboard.dismiss();
  };

  const handleToggleCompleteTask = ({ id, completed }) => {
    const copy = { ...tasks };
    const task = copy[selectedDate].find(t => t.id === id);
    task.completed = completed;
    setTasks(copy);
  };

  const handleRemoveTask = id => {
    const newData = { ...tasks };
    const prevIndex = newData[selectedDate].findIndex(item => item.id === id);
    newData[selectedDate].splice(prevIndex, 1);
    setTasks(newData);
  };

  const handleToggleCalendar = () => {
    if (taskRowBeingMoved) taskRowBeingMoved.closeRow();
    setTaskRowBeingMoved(null);

    return toggleCalendarVisibility(!calendarVisible);
  };

  const handleMoveTask = date => {
    const { dateString } = date;
    const { completed, description, id } = taskRowBeingMoved.props.item;
    const taskToBeMoved = {
      completed,
      description,
      id: String(Date.now())
    };
    const newData = { ...tasks };
    if (newData[dateString]) {
      newData[dateString].unshift(taskToBeMoved);
    } else {
      newData[dateString] = [taskToBeMoved];
    }
    const prevIndex = newData[selectedDate].findIndex(item => item.id === id);
    newData[selectedDate].splice(prevIndex, 1);
    setTasks(newData);
    setTaskRowBeingMoved(null);
  };

  return (
    <KeyboardAvoidingView
      style={{
        height: "100%",
        display: "flex",
        flex: 1,
        backgroundColor: "#F7F8FA"
      }}
      contentContainerStyle={{ height: "100%", display: "flex", flex: 1 }}
      behavior="position"
      enabled
      keyboardVerticalOffset={calendarVisible ? -200 : -600}
    >
      <ScrollView
        contentContainerStyle={styles.container}
        keyboardShouldPersistTaps="handled"
      >
        <View style={styles.innerMainScreenContainer}>
          <CurrentDayButton
            selectedDate={selectedDate}
            handlePress={handleToggleCalendar}
            buttonPressed={!calendarVisible}
          />
          {calendarVisible && (
            <View style={styles.calendarContainer}>
              <Calendar
                selectedDate={selectedDate}
                handleDayPress={handleDayPress}
              />
            </View>
          )}
          <TaskInput handleAddTask={handleAddTask} />
          {tasks[selectedDate] && (
            <SafeAreaView>
              <SwipeableTaskList
                listViewData={tasks[selectedDate]
                  .sort((a, b) => a.completed - b.completed)
                  .map(task => ({
                    key: task.id,
                    ...task
                  }))}
                dimensions={{
                  width: Dimensions.get("window").width,
                  height: Dimensions.get("window").height
                }}
                handleToggleCompleteTask={handleToggleCompleteTask}
                handleRemoveTask={handleRemoveTask}
                swipeDisabled={calendarVisible}
                handleRenderMoveMap={row => setTaskRowBeingMoved(row)}
              />
            </SafeAreaView>
          )}
        </View>
        {taskRowBeingMoved && (
          <View style={styles.moveTaskCalendarContainer}>
            <Calendar
              calendarBackgroundColor="white"
              selectedDate={selectedDate}
              handleDayPress={handleMoveTask}
              calendarLabel="Start Date:"
            />
          </View>
        )}
      </ScrollView>
    </KeyboardAvoidingView>
  );
}

const styles = StyleSheet.create({
  standalone: {
    marginTop: 30,
    marginBottom: 30
  },
  innerMainScreenContainer: {
    paddingLeft: 24,
    paddingRight: 24
  },
  container: {
    paddingTop: 48,
    paddingBottom: 48,
    width: "100%",
    backgroundColor: "#F7F8FA"
  },
  moveTaskCalendarContainer: {
    width: "100%",
    display: "flex",
    flexDirection: "column",
    flex: 1,
    alignItems: "center",
    backgroundColor: "white"
  },
  calendarContainer: {
    width: "100%",
    display: "flex",
    flexDirection: "column",
    flex: 1,
    alignItems: "center"
  }
});
