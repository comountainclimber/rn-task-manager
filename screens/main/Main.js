import React from "react";
import {
  StyleSheet,
  ScrollView,
  KeyboardAvoidingView,
  Dimensions,
  View
} from "react-native";
import moment from "moment";
import PropTypes from "prop-types";

import { DATE } from "../../constants";
import Calendar from "../../components/Calendar";
import CurrentDayButton from "../../components/CurrentDayButton";
import TaskInput from "../../components/TaskInput";
import SwipeableTaskList from "../../components/SwipeableTaskList";

export default function Main({
  tasks,
  handleAddTask,
  handleToggleCompleteTask,
  handleRemoveTask,
  handleMoveTask
}) {
  const [selectedDate, setSelectedDate] = React.useState(
    moment(DATE).format("YYYY-MM-DD")
  );
  const [calendarVisible, toggleCalendarVisibility] = React.useState(false);
  const [taskRowBeingMoved, setTaskRowBeingMoved] = React.useState(null);

  const handleDayPress = date => setSelectedDate(date.dateString);

  const handleToggleCalendar = () => {
    if (taskRowBeingMoved) taskRowBeingMoved.closeRow();
    setTaskRowBeingMoved(null);
    return toggleCalendarVisibility(!calendarVisible);
  };

  const dimensions = {
    width: Dimensions.get("window").width,
    height: Dimensions.get("window").height
  };

  return (
    <KeyboardAvoidingView
      style={{
        height: "100%",
        display: "flex",
        flex: 1,
        backgroundColor: "#F7F8FA"
      }}
      contentContainerStyle={{ height: "100%", display: "flex", flex: 1 }}
      behavior="position"
      enabled
      keyboardVerticalOffset={calendarVisible ? -200 : -600}
    >
      <ScrollView
        contentContainerStyle={styles.container}
        keyboardShouldPersistTaps="handled"
      >
        <View
          onStartShouldSetResponder={() =>
            taskRowBeingMoved.closeRow() && setTaskRowBeingMoved(null)
          }
          style={styles.innerMainScreenContainer}
        >
          <CurrentDayButton
            selectedDate={selectedDate}
            handlePress={handleToggleCalendar}
            buttonPressed={!calendarVisible}
          />
          {calendarVisible && (
            <View style={styles.calendarContainer}>
              <Calendar
                selectedDate={selectedDate}
                handleDayPress={handleDayPress}
                dimensions={dimensions}
              />
            </View>
          )}
          <TaskInput
            handleAddTask={({ description }) =>
              handleAddTask({ task: { description }, selectedDate })
            }
          />
          {tasks[selectedDate] && (
            <SwipeableTaskList
              listViewData={tasks[selectedDate]
                .sort((a, b) => a.completed - b.completed)
                .map(task => ({
                  key: task.id,
                  ...task
                }))}
              dimensions={dimensions}
              handleToggleCompleteTask={({ id, completed }) => {
                return handleToggleCompleteTask({
                  task: {
                    ...tasks[selectedDate].find(t => t.id === id),
                    completed
                  },
                  selectedDate
                });
              }}
              handleRemoveTask={id => handleRemoveTask({ id, selectedDate })}
              swipeDisabled={calendarVisible}
              handleRenderMoveMap={row => setTaskRowBeingMoved(row)}
            />
          )}
        </View>
        {taskRowBeingMoved && (
          <View style={styles.moveTaskCalendarContainer}>
            <Calendar
              dimensions={dimensions}
              calendarBackgroundColor="white"
              selectedDate={selectedDate}
              handleDayPress={date => {
                const { dateString } = date;
                const {
                  completed,
                  description,
                  id
                } = taskRowBeingMoved.props.item;
                const task = {
                  completed,
                  description,
                  id
                };
                setTaskRowBeingMoved(null);
                return handleMoveTask({
                  task,
                  selectedDate: dateString,
                  previousDate: selectedDate
                });
              }}
              calendarLabel="Start Date:"
            />
          </View>
        )}
      </ScrollView>
    </KeyboardAvoidingView>
  );
}

Main.propTypes = {
  tasks: PropTypes.shape({
    date: PropTypes.arrayOf(
      PropTypes.exact({
        id: PropTypes.string,
        description: PropTypes.string,
        completed: PropTypes.bool
      })
    )
  }).isRequired,
  handleAddTask: PropTypes.func.isRequired,
  handleRemoveTask: PropTypes.func.isRequired,
  handleMoveTask: PropTypes.func.isRequired,
  handleToggleCompleteTask: PropTypes.func.isRequired
};

const styles = StyleSheet.create({
  standalone: {
    marginTop: 30,
    marginBottom: 30
  },
  innerMainScreenContainer: {
    paddingLeft: 24,
    paddingRight: 24
  },
  container: {
    paddingTop: 48,
    paddingBottom: 48,
    width: "100%",
    backgroundColor: "#F7F8FA"
  },
  moveTaskCalendarContainer: {
    display: "flex",
    flexDirection: "column",
    flex: 1,
    alignItems: "center",
    backgroundColor: "white"
  },
  calendarContainer: {
    width: "100%",
    display: "flex",
    flexDirection: "column",
    flex: 1,
    alignItems: "center"
  }
});
