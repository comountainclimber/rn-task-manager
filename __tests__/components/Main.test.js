import React from "react";
import renderer from "react-test-renderer";

import Main from "../../screens/main/Main";

const mock = () => undefined;

it("it renders without crashing", () => {
  const tree = renderer
    .create(
      <Main
        tasks={{}}
        handleToggleCompleteTask={mock}
        handleMoveTask={mock}
        handleAddTask={mock}
        handleRemoveTask={mock}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
