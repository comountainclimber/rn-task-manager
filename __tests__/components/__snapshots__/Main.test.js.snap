// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders without crashing 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "backgroundColor": "#F7F8FA",
      "display": "flex",
      "flex": 1,
      "height": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flex": 1,
          "height": "100%",
        },
        Object {
          "bottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        Object {
          "backgroundColor": "#F7F8FA",
          "paddingBottom": 48,
          "paddingTop": 48,
          "width": "100%",
        }
      }
      keyboardShouldPersistTaps="handled"
    >
      <View>
        <View
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "paddingLeft": 24,
              "paddingRight": 24,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "marginBottom": 24,
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "column",
                  "marginRight": 12,
                }
              }
            >
              <Text
                style={
                  Object {
                    "fontFamily": "Nunito-Bold",
                    "fontSize": 24,
                    "textAlign": "left",
                    "width": "100%",
                  }
                }
              >
                Tuesday
              </Text>
              <Text
                style={
                  Object {
                    "color": "grey",
                    "fontFamily": "Nunito-Regular",
                    "fontSize": 16,
                    "marginTop": -2,
                    "textAlign": "left",
                    "width": "100%",
                  }
                }
              >
                February 25
              </Text>
            </View>
            <test-file-stub />
          </View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#F0F0F0",
                  "borderRadius": 4,
                  "borderWidth": 1.5,
                  "marginBottom": 12,
                  "padding": 12,
                  "width": "100%",
                },
                Object {
                  "padding": 12,
                },
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              blurOnSubmit={true}
              multiline={true}
              onChangeText={[Function]}
              onSubmitEditing={[Function]}
              placeholder="Add a task"
              rejectResponderTermination={true}
              style={
                Object {
                  "fontFamily": "Nunito-Regular",
                }
              }
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
